src/linear_algebra/Hermitian_inner_product.lean

```diff
@@ -0,0 +1,669 @@
+/-
+Copyright (c) 2018 Andreas Swerdlow. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Andreas Swerdlow
+-/
+
+import analysis.normed_space.basic linear_algebra.sesquilinear_form data.complex.basic
+
+open vector_space field set complex real
+
+set_option class.instance_max_depth 32
+
+lemma ne_comm {α : Type*} {a b : α} : a ≠ b ↔ b ≠ a :=
+⟨ λ H, iff_subst (@eq_comm _ a b) H, 
+  λ H, iff_subst (@eq_comm _ b a) H⟩ 
+
+lemma im_eq_zero_iff_conj_eq (x : ℂ) : x.im = 0 ↔ conj(x) = x :=
+begin
+  dunfold conj,
+  split; intro H,
+  { rw [H, neg_zero, ←H, complex.eta] },
+
+  { rw complex.ext_iff at H,
+    have : -x.im = x.im,
+    { exact H.right },
+    rw ←add_eq_zero_iff_neg_eq at this,
+    exact add_self_eq_zero.mp this}
+end
+
+lemma ne_zero_im_zero_imp_re_ne_zero {x : ℂ} : x ≠ 0 → x.im = 0 → x.re ≠ 0 :=
+begin
+  intros H1 H2,
+  have Hx : x = ↑x.re,
+  { rw [←re_add_im x, H2, of_real_zero, 
+        zero_mul, field.add_zero, of_real_re] },
+  rw Hx at H1,
+  exact of_real_ne_zero.mp H1,
+end
+
+lemma re_of_real {x : ℂ} : x.im = 0 → ↑(x.re) = x :=
+begin
+  intros H,
+  rw [←re_add_im x, H, of_real_zero, zero_mul, field.add_zero, of_real_inj, of_real_re],
+end
+
+lemma of_real_pow (x : ℝ) (a : ℕ) : (↑(x^a) : ℂ) = (↑x)^a :=
+begin
+  induction a with d Hd,
+  { simp },
+  { rw [pow_succ, pow_succ],
+    rw of_real_mul,
+    rw Hd },
+end
+
+def conj.equiv : equiv ℂ ℂ := 
+⟨conj, conj, begin dunfold function.left_inverse, intros, simp, end, begin dunfold function.right_inverse, dunfold function.left_inverse, intros, simp, end⟩
+
+noncomputable def conj.ring_equiv : ℂ ≃r ℂ := 
+⟨conj.equiv, conj_one, conj_mul, conj_add⟩ 
+
+noncomputable def conj.ring_invo : ring_invo ℂ :=
+⟨comm_ring.isom_to_anti_isom conj.ring_equiv, begin apply conj_conj end⟩
+
+class herm_inner_product_space (α : Type*) [add_comm_group α] [vector_space ℂ α] extends sym_sesq_form ℂ α conj.ring_invo :=
+(sesq_self_re_nonneg : ∀ (x : α), (sesq x x).re ≥ 0)
+(sesq_self_eq_zero_iff : ∀ (x : α), sesq x x = 0 ↔ x = 0)
+
+namespace herm_inner_product_space
+
+open herm_inner_product_space ring_invo
+
+variables {α : Type*} [add_comm_group α] [vector_space ℂ α] [herm_inner_product_space α]
+
+noncomputable def inprod : α → α → ℂ := (herm_inner_product_space.to_sym_sesq_form α).to_sesq_form.sesq  
+
+infix `₀` : 74 := inprod 
+
+lemma conj_inprod (x y : α) : conj (x ₀ y) = y ₀ x := sym_sesq_form.map_sesq conj.ring_invo y x 
```

Comment:
The following lemmas should all be put in the `inprod` (or `inner_product`, see above) namespace.